java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java

```diff
@@ -38,21 +39,46 @@
 import org.sonar.plugins.java.api.tree.Tree;
 import org.sonar.plugins.java.api.tree.Tree.Kind;
 
+import javax.annotation.Nullable;
+
+import java.util.HashMap;
 import java.util.List;
-import java.util.function.Predicate;
+import java.util.Map;
 import java.util.stream.Stream;
 
 @Rule(key = "S2187")
 public class NoTestInTestClassCheck extends IssuableSubscriptionVisitor {
 
-  private static final Predicate<SymbolMetadata.AnnotationInstance> PREDICATE_ANNOTATION_TEST_OR_UNKNOWN = input -> {
-    Type type = input.symbol().type();
-    return type.isUnknown() || isTestAnnotation(type);
-  };
+  private final Map<String, Boolean> testAnnotations;
+
+  public NoTestInTestClassCheck() {
+    testAnnotations = new HashMap<>();
```

Comment:
This map should probably be replaced by two sets : "encouteredAnnotations" and "testableAnnotations" : this should make the code simpler in the end.